query/src/users/auth/auth_mgr.rs

```diff
@@ -42,28 +42,37 @@ pub enum Credential {
 }
 
 impl AuthMgr {
-    pub async fn create(cfg: Config, users: Arc<UserApiProvider>) -> Result<Self> {
+    pub async fn create(cfg: Config, user_mgr: Arc<UserApiProvider>) -> Result<Self> {
         Ok(AuthMgr {
-            users,
+            user_mgr,
             tenant: cfg.query.tenant_id.clone(),
             jwt: JwtAuthenticator::try_create(cfg).await?,
         })
     }
 
     pub async fn no_auth(&self) -> Result<UserInfo> {
-        self.users
+        self.user_mgr
             .get_user(&self.tenant, UserIdentity::new("root", "127.0.0.1"))
             .await
     }
 
     pub async fn auth(&self, credential: &Credential) -> Result<UserInfo> {
         match credential {
             Credential::Jwt { token: t } => {
-                let user_name = match &self.jwt {
-                    Some(j) => j.get_user(t.as_str())?,
+                let jwt = match &self.jwt {
+                    Some(j) => j.get_jwt_claims(t.as_str())?,
                     None => return Err(ErrorCode::AuthenticateFailure("jwt auth not configured.")),
                 };
-                self.users
+                let user_name = jwt.subject.unwrap();
+                if let Some(create_user) = jwt.custom.create_user {
```

Comment:
1. `add_user`  return error if user exists. does the cloud platform when it should carry the create_user field?

2. for an existed user,  what  if  the existed user info diff with info in create_user?

Reply:
> add_user return error if user exists. does the cloud platform when it should carry the create_user field?

User will be created only if user does not exist. `is_not_exists` is true for `add_user `.

> for an existed user, what if the existed user info diff with info in create_user?

`create_user` just used to create users, not update.

Reply:
> User will be created only if user does not exist

do you mean  cloud platform do this check  to determine whether or not to add`create_user` in jwt?  cc @flaneur2020   


`add_user` may ErrorCode::UserAlreadyExists

Reply:
> do you mean cloud platform do this check to determine whether or not to addcreate_user in jwt?

i guess the cloud platform could ensure user exists by an jwt with create_user enabled when the user successfully logged in, and then pass normal jwt for the normal operations to interact with databend-query.

a jwt is used to authenticate an user, if the jwt verified successfully and the user already existed, it can pass the authentication happily, it might a bit weird to get a userAlreadyExists in this case. 🤔

Reply:
btw, how about making the `CreateUser` claim into a bool?

the infomation about tenant and role may already existed some another claims 🤔

Reply:
> `add_user` may ErrorCode::UserAlreadyExists

@youngsofun `UserAlreadyExists` will be ignored if the user already exists by `is_not_exists` is true.

Reply:
> btw, how about making the create_user claim into a bool?

like this?
```
{
    "create_user": true,
    "tenant_id": "",
    "roles": []
}
```

I'm not sure if the same fields would be in custom claims, it's better together.

Reply:
@Junnplus sorry, what I  saw is UserApi::add_user.

Reply:
> i guess the cloud platform could ensure user exists by an jwt with create_user enabled when the user successfully logged in, and then pass normal jwt for the normal operations to interact with databend-query.


maybe we can call it  ```ensure_user```?  I saw it a lot in puppet.  or "create_user_if_not_exist".

> I'm not sure if the same fields would be in custom claims, it's better together

seems currently this `create_user` provides only a `basic` template for a newly-seen user, and it will not override behavior specified by an existing userInfo, I prefer separate from other fields (which may later be used to override some behavior of the session)

Reply:
`ensure_user` SGTM.

Reply:
cool, `ensure_user` is much better than `create_user_if_not_exist`.

I guess jwt in some manner favors shorter names, as it'd like to avoids getting too big in the transport, likewise "aud" over "audience", "sub" over "subject" etc.